File: web/src/privateCallStabilizer.js
import { ConnectionState, DisconnectReason, RoomEvent } from 'livekit-client';
import { fetchPrivateCall, pingPrivateCall, refreshPrivateCall } from './api';
import { publicLiveKitUrl } from './livekitUrl';

export const CALL_PEER_GRACE_MS = 45_000;
export const CALL_PING_MS = 25_000;
export const CALL_RECONNECT_DELAYS = [3_000, 6_000, 12_000, 20_000];
export const CALL_MAX_PEER_GRACE_CYCLES = 5;
export const CALL_MAX_RECONNECT_CYCLES = 6;

function isCallGoneError(err) {
  const msg = String(err?.message || err || '').toLowerCase();
  const status = err?.status || err?.statusCode;
  return (
    status === 404 ||
    msg.includes('404') ||
    msg.includes('no encontrada') ||
    msg.includes('not found') ||
    msg.includes('finalizada') ||
    msg.includes('ended')
  );
}

/**
 * Estabilizador suave: LiveKit reconecta solo; nosotros solo intervenimos
 * si queda Disconnected de verdad. Tras N ciclos de grace/reconnect → onGiveUp.
 */
export function attachPrivateCallStabilizer({
  room,
  callId,
  authToken,
  signalSocket = null,
  closingRef,
  onStatus,
  onRemoteEnd,
  onGiveUp,
  getPeerLabel = () => 'el otro usuario',
  peerGraceMs = CALL_PEER_GRACE_MS,
  pingMs = CALL_PING_MS,
  maxPeerGraceCycles = CALL_MAX_PEER_GRACE_CYCLES,
  maxReconnectCycles = CALL_MAX_RECONNECT_CYCLES,
}) {
  if (!room || !callId || !authToken) return () => {};

  let peerGraceTimer = null;
  let pingTimer = null;
  let reconnectAttempt = 0;
  let peerGraceCycles = 0;
  let reconnectCycles = 0;
  let reconnectTimer = null;
  let disposed = false;
  let quietStatusUntil = 0;

  const clearPeerGrace = () => {
    if (peerGraceTimer) {
      clearTimeout(peerGraceTimer);
      peerGraceTimer = null;
    }
  };

  const busy = () =>
    room.state === ConnectionState.Connected ||
    room.state === ConnectionState.Reconnecting ||
    room.state === ConnectionState.Connecting ||
    room.state === ConnectionState.SignalReconnecting;

  const abandon = async (status) => {
    if (disposed || closingRef?.current) return;
    onStatus?.(status);
    const fn = onGiveUp || onRemoteEnd;
    await fn?.();
  };

  const verifyCallActive = async () => {
    try {
      const data = await fetchPrivateCall(authToken, callId);
      const status = data?.call?.status;
      return Boolean(status && status !== 'ended');
    } catch (err) {
      if (isCallGoneError(err)) return false;
      // Blip de API: no dar por muerta.
      return true;
    }
  };

  const schedulePeerGrace = () => {
    if (disposed || closingRef?.current) return;
    clearPeerGrace();
    peerGraceTimer = setTimeout(() => {
      if (disposed || closingRef?.current) return;
      if (Date.now() < quietStatusUntil) return;
      onStatus?.(`Reconectando con ${getPeerLabel()}…`);
      peerGraceTimer = setTimeout(async () => {
        if (disposed || closingRef?.current) return;
        const alive = await verifyCallActive();
        if (!alive) {
          onRemoteEnd?.();
          return;
        }
        peerGraceCycles += 1;
        if (peerGraceCycles >= maxPeerGraceCycles) {
          await abandon('Sin respuesta del otro usuario');
          return;
        }
        onStatus?.('Conexión inestable — manteniendo llamada…');
        schedulePeerGrace();
      }, peerGraceMs);
    }, 8000);
  };

  const scheduleReconnect = () => {
    if (disposed || closingRef?.current || busy()) return;
    if (reconnectCycles >= maxReconnectCycles) {
      abandon('No se pudo restaurar la conexión');
      return;
    }
    if (reconnectAttempt >= CALL_RECONNECT_DELAYS.length) {
      reconnectAttempt = 0;
      reconnectCycles += 1;
    }
    const delay = CALL_RECONNECT_DELAYS[reconnectAttempt] ?? 20_000;
    reconnectAttempt += 1;
    clearTimeout(reconnectTimer);
    reconnectTimer = setTimeout(async () => {
      if (disposed || closingRef?.current || busy()) return;
      const alive = await verifyCallActive();
      if (!alive) {
        onRemoteEnd?.();
        return;
      }
      try {
        onStatus?.('Restaurando enlace…');
        const fresh = await refreshPrivateCall(authToken, callId);
        const url = publicLiveKitUrl(fresh.url);
        const token = fresh.token;
        if (busy()) return;
        await room.connect(url, token);
        reconnectAttempt = 0;
        reconnectCycles = 0;
        quietStatusUntil = Date.now() + 5000;
        onStatus?.('Enlace restaurado');
      } catch (err) {
        if (isCallGoneError(err)) {
          onRemoteEnd?.();
          return;
        }
        scheduleReconnect();
      }
    }, delay);
  };

  const resyncFromBackend = async () => {
    if (disposed || closingRef?.current) return;
    try {
      await pingPrivateCall(authToken, callId);
    } catch (err) {
      if (isCallGoneError(err)) {
        onRemoteEnd?.();
        return;
      }
    }
    if (room.state === ConnectionState.Disconnected) {
      scheduleReconnect();
    }
  };

  const onParticipantDisconnected = () => {
    if (disposed || closingRef?.current) return;
    schedulePeerGrace();
  };

  const onParticipantConnected = () => {
    if (disposed || closingRef?.current) return;
    clearPeerGrace();
    reconnectAttempt = 0;
    peerGraceCycles = 0;
    reconnectCycles = 0;
    quietStatusUntil = Date.now() + 4000;
  };

  const onReconnecting = () => {
    if (disposed || closingRef?.current) return;
    if (Date.now() < quietStatusUntil) return;
    onStatus?.('Reconectando…');
  };

  const onReconnected = () => {
    if (disposed || closingRef?.current) return;
    clearPeerGrace();
    clearTimeout(reconnectTimer);
    reconnectAttempt = 0;
    peerGraceCycles = 0;
    reconnectCycles = 0;
    quietStatusUntil = Date.now() + 6000;
  };

  const onDisconnected = (reason) => {
    if (disposed || closingRef?.current) return;
    if (
      reason === DisconnectReason.CLIENT_INITIATED ||
      reason === DisconnectReason.ROOM_DELETED ||
      reason === DisconnectReason.DUPLICATE_IDENTITY
    ) {
      return;
    }
    clearTimeout(reconnectTimer);
    reconnectTimer = setTimeout(() => {
      if (disposed || closingRef?.current) return;
      if (room.state === ConnectionState.Disconnected) scheduleReconnect();
    }, 2500);
  };

  room.on(RoomEvent.ParticipantDisconnected, onParticipantDisconnected);
  room.on(RoomEvent.ParticipantConnected, onParticipantConnected);
  room.on(RoomEvent.Reconnecting, onReconnecting);
  room.on(RoomEvent.Reconnected, onReconnected);
  room.on(RoomEvent.Disconnected, onDisconnected);

  pingTimer = setInterval(() => {
    if (disposed || closingRef?.current) return;
    pingPrivateCall(authToken, callId).catch((err) => {
      if (isCallGoneError(err)) onRemoteEnd?.();
    });
  }, pingMs);

  const onSocketConnect = () => {
    quietStatusUntil = Date.now() + 3000;
    resyncFromBackend();
  };
  if (signalSocket) {
    signalSocket.on('connect', onSocketConnect);
    signalSocket.on('reconnect', onSocketConnect);
  }

  return () => {
    disposed = true;
    clearPeerGrace();
    clearInterval(pingTimer);
    clearTimeout(reconnectTimer);
    room.off(RoomEvent.ParticipantDisconnected, onParticipantDisconnected);
    room.off(RoomEvent.ParticipantConnected, onParticipantConnected);
    room.off(RoomEvent.Reconnecting, onReconnecting);
    room.off(RoomEvent.Reconnected, onReconnected);
    room.off(RoomEvent.Disconnected, onDisconnected);
    if (signalSocket) {
      signalSocket.off('connect', onSocketConnect);
      signalSocket.off('reconnect', onSocketConnect);
    }
  };
}

import { mergeStreamingRoomOptions } from './videoStreaming.js';

export const RESILIENT_ROOM_OPTIONS = mergeStreamingRoomOptions();

File: mobile/lib/private_call_stabilizer.dart
import 'dart:async';

import 'package:livekit_client/livekit_client.dart';
import 'package:socket_io_client/socket_io_client.dart' as io;

import 'api_client.dart';
import 'config.dart';

/// Estabilizador suave: deja que LiveKit reconecte; no cuelga por microcortes.
/// Tras N ciclos de grace/reconnect, abandona (onGiveUp / onRemoteEnd).
class PrivateCallStabilizer {
  PrivateCallStabilizer({
    required this.api,
    required this.callId,
    required this.room,
    required this.liveKitUrl,
    required this.liveKitToken,
    required this.onStatus,
    required this.onRemoteEnd,
    required this.isClosing,
    this.onGiveUp,
    this.peerLabel = 'el otro usuario',
    this.peerGrace = const Duration(seconds: 45),
    this.maxPeerGraceCycles = 5,
    this.maxReconnectCycles = 6,
  });

  final ApiClient api;
  final String callId;
  final Room room;
  final String liveKitUrl;
  final String liveKitToken;
  final void Function(String status) onStatus;
  final Future<void> Function() onRemoteEnd;
  /// Si se agotan reintentos: cuelga en servidor + cierra UI.
  final Future<void> Function()? onGiveUp;
  final bool Function() isClosing;
  final String peerLabel;
  final Duration peerGrace;
  final int maxPeerGraceCycles;
  final int maxReconnectCycles;

  static const _pingEvery = Duration(seconds: 25);
  static const _reconnectDelays = [
    Duration(seconds: 3),
    Duration(seconds: 6),
    Duration(seconds: 12),
    Duration(seconds: 20),
  ];

  Timer? _peerGraceTimer;
  Timer? _pingTimer;
  Timer? _reconnectTimer;
  int _reconnectAttempt = 0;
  int _peerGraceCycles = 0;
  int _reconnectCycles = 0;
  EventsListener<RoomEvent>? _listener;
  bool _disposed = false;
  DateTime _quietUntil = DateTime.fromMillisecondsSinceEpoch(0);

  bool get _busy =>
      room.connectionState == ConnectionState.connected ||
      room.connectionState == ConnectionState.reconnecting ||
      room.connectionState == ConnectionState.connecting;

  Future<void> _abandon(String status) async {
    if (_disposed || isClosing()) return;
    onStatus(status);
    final fn = onGiveUp ?? onRemoteEnd;
    await fn();
  }

  void attach({io.Socket? signalSocket}) {
    _pingTimer = Timer.periodic(_pingEvery, (_) {
      if (_disposed || isClosing()) return;
      unawaited(_pingOrDie());
    });

    signalSocket?.onConnect((_) => _resyncFromBackend());
    signalSocket?.onReconnect((_) => _resyncFromBackend());

    final listener = room.createListener();
    _listener = listener;

    listener.on<ParticipantDisconnectedEvent>((_) => _schedulePeerGrace());
    listener.on<ParticipantConnectedEvent>((_) {
      _cancelPeerGrace();
      _reconnectAttempt = 0;
      _peerGraceCycles = 0;
      _reconnectCycles = 0;
      _quietUntil = DateTime.now().add(const Duration(seconds: 4));
    });
    listener.on<RoomReconnectingEvent>((_) {
      if (DateTime.now().isBefore(_quietUntil)) return;
      onStatus('Reconectando…');
    });
    listener.on<RoomReconnectedEvent>((_) {
      _cancelPeerGrace();
      _reconnectTimer?.cancel();
      _reconnectAttempt = 0;
      _peerGraceCycles = 0;
      _reconnectCycles = 0;
      _quietUntil = DateTime.now().add(const Duration(seconds: 6));
    });
    listener.on<RoomDisconnectedEvent>((e) {
      if (e.reason == DisconnectReason.clientInitiated ||
          e.reason == DisconnectReason.roomDeleted ||
          e.reason == DisconnectReason.duplicateIdentity) {
        return;
      }
      _reconnectTimer?.cancel();
      _reconnectTimer = Timer(const Duration(milliseconds: 2500), () {
        if (_disposed || isClosing()) return;
        if (room.connectionState == ConnectionState.disconnected) {
          _scheduleReconnect();
        }
      });
    });
  }

  Future<void> _pingOrDie() async {
    try {
      await api.pingPrivateCall(callId);
    } catch (e) {
      final msg = e.toString().toLowerCase();
      if (msg.contains('404') ||
          msg.contains('no encontrada') ||
          msg.contains('not found') ||
          msg.contains('finalizada') ||
          msg.contains('ended')) {
        await onRemoteEnd();
      }
    }
  }

  /// true = sigue activa; false = terminada en servidor.
  Future<bool> _verifyCallActive() async {
    try {
      final data = await api.fetchPrivateCall(callId);
      final call = data['call'];
      if (call is! Map) return false;
      final status = call['status']?.toString();
      return status != null && status != 'ended';
    } catch (e) {
      final msg = e.toString().toLowerCase();
      if (msg.contains('404') ||
          msg.contains('no encontrada') ||
          msg.contains('not found') ||
          msg.contains('finalizada') ||
          msg.contains('ended')) {
        return false;
      }
      // Blip de red: no dar por muerta.
      return true;
    }
  }

  void _cancelPeerGrace() {
    _peerGraceTimer?.cancel();
    _peerGraceTimer = null;
  }

  void _schedulePeerGrace() {
    if (_disposed || isClosing()) return;
    _cancelPeerGrace();
    _peerGraceTimer = Timer(const Duration(seconds: 8), () {
      if (_disposed || isClosing()) return;
      if (DateTime.now().isBefore(_quietUntil)) return;
      onStatus('Reconectando con $peerLabel…');
      _peerGraceTimer = Timer(peerGrace, () async {
        if (_disposed || isClosing()) return;
        final alive = await _verifyCallActive();
        if (!alive) {
          await onRemoteEnd();
          return;
        }
        _peerGraceCycles += 1;
        if (_peerGraceCycles >= maxPeerGraceCycles) {
          await _abandon('Sin respuesta del otro usuario');
          return;
        }
        onStatus('Conexión inestable — manteniendo llamada…');
        _schedulePeerGrace();
      });
    });
  }

  void _scheduleReconnect() {
    if (_disposed || isClosing() || _busy) return;
    if (_reconnectCycles >= maxReconnectCycles) {
      unawaited(_abandon('No se pudo restaurar la conexión'));
      return;
    }
    if (_reconnectAttempt >= _reconnectDelays.length) {
      _reconnectAttempt = 0;
      _reconnectCycles += 1;
    }
    final delay = _reconnectDelays[_reconnectAttempt];
    _reconnectAttempt += 1;
    _reconnectTimer?.cancel();
    _reconnectTimer = Timer(delay, () async {
      if (_disposed || isClosing() || _busy) return;
      final alive = await _verifyCallActive();
      if (!alive) {
        await onRemoteEnd();
        return;
      }
      try {
        onStatus('Restaurando enlace…');
        final fresh = await api.refreshPrivateCall(callId);
        final url = AppConfig.publicLiveKitUrl(fresh['url']?.toString() ?? liveKitUrl);
        final token = fresh['token']?.toString() ?? liveKitToken;
        if (_busy) return;
        await room.connect(url, token);
        _reconnectAttempt = 0;
        _reconnectCycles = 0;
        _quietUntil = DateTime.now().add(const Duration(seconds: 5));
        onStatus('Enlace restaurado');
      } catch (e) {
        final msg = e.toString().toLowerCase();
        if (msg.contains('404') || msg.contains('no encontrada') || msg.contains('not found')) {
          await onRemoteEnd();
          return;
        }
        _scheduleReconnect();
      }
    });
  }

  Future<void> _resyncFromBackend() async {
    if (_disposed || isClosing()) return;
    _quietUntil = DateTime.now().add(const Duration(seconds: 3));
    try {
      await api.pingPrivateCall(callId);
    } catch (e) {
      final msg = e.toString().toLowerCase();
      if (msg.contains('404') ||
          msg.contains('no encontrada') ||
          msg.contains('not found') ||
          msg.contains('finalizada')) {
        await onRemoteEnd();
        return;
      }
    }
    if (room.connectionState == ConnectionState.disconnected) {
      _scheduleReconnect();
    }
  }

  void dispose() {
    _disposed = true;
    _cancelPeerGrace();
    _pingTimer?.cancel();
    _reconnectTimer?.cancel();
    _listener?.dispose();
    _listener = null;
  }
}
